include/gz/math/TimeVaryingVolumetricGridLookupField.hh

```diff
@@ -0,0 +1,291 @@
+          res.push_back(slice2);
+        }
+        return res;
+      }
+
+      /// \brief Uses quadrilinear interpolation to estimate value of current
+      /// point. Returns nullopt if query is out of range.
+      /// \param _session - The session
+      /// \param _interpolators - The interpolation points
+      /// retrieved from `LookUp()`
+      /// \param _values1 - Value array at timestep 1.
+      /// \param _values2 - Value array at timestep 2.
+      /// \param _position - The position to be queried.
+      /// \param _default - Value used if there is a hole in the data.
+      /// \returns The estimated value for the point. Nullopt if we are
+      /// outside the field. Default value if in the field but no value is
+      /// in the index.
+      public: template<typename X>
+      std::optional<X> EstimateQuadrilinear(
+        const InMemorySession<T, V> &_session,
+        const std::vector<InterpolationPoint4D<T, V>> &_interpolators,
+        const std::vector<X> &_values1,
+        const std::vector<X> &_values2,
+        const Vector3<X> &_position,
+        const X _default = X(0)
+      ) const
+      {
+        if (_session.iter == this->gridFields.end())
+        {
+          // Out of bounds
+          return std::nullopt;
+        }
+
+        auto time = _session.time;
+        if (_interpolators.size() == 0) return std::nullopt;
+        if (_interpolators.size() == 1)
+        {
+          // This happens we reach the end of time
+          return _session.iter->second.EstimateValueUsingTrilinear(
+            _position,
+            _values2,
+            _default);
+        }
+
+        /// Got nothing to interpolate. Out of bounds.
+        if (_interpolators[0].timeSlice.size() == 0
+         && _interpolators[1].timeSlice.size() == 0)
+          return std::nullopt;
+
+        /// Only one of the two time-slices has data. Use that slice to guess.
+        auto next = std::next(_session.iter);
+        if (_interpolators[1].timeSlice.size() == 0)
+        {
+          return _session.iter->second.EstimateValueUsingTrilinear(
+            _position,
+            _values1,
+            _default
+          );
+        }
+        if (_interpolators[0].timeSlice.size() == 0)
+        {
+          return next->second.EstimateValueUsingTrilinear(
+            _position,
+            _values2,
+            _default
+          );
+        }
+
+        /// Default case where both time-slices has interpolation
+        auto res1 = _session.iter->second.EstimateValueUsingTrilinear(
+          _position,
+          _values1,
+          _default
+        );
+
+        auto res2 = next->second.EstimateValueUsingTrilinear(
+          _position,
+          _values2,
+          _default
+        );
+
+        if (res1.has_value() || res2.has_value())
+        {
+          InterpolationPoint1D<T>
+            pt1{_session.iter->first, 0}, pt2{next->first, 1};
+          // If either has value interpolate using default value.
```

Comment:
@arjo129 meta: ok with this behavior for a first pass, but it may not be appropriate for several use cases. Perhaps we can add a ticket to generalize this a bit (e.g. to allow constant extrapolation or simply return nullopt or whatever you can think of).